#### Université de Bordeaux,  Master Mention Informatique

# Analyse, classification et indexation des données: feuille 2
### Introduction : métriques d'évaluation

### Présentation

L'objectif de ce TD est de comprendre les métriques d'évaluation des modèles de machine learning. 

### Exercice 1.
On dispose d'un corpus de données $D = (x_i, y_i)_{1\leq i \leq n}$. On partitionne alors $D$ en deux ensembles :
$$
D = Train \cup Test
$$
Le sous ensemble $Train$ est alors utilisé pour entrainer un modèle $M$ alors que $Test$ est lui utilisé pour tester $M$

1. Rappeler la définition de la matrice de confusion 
2. On se limite au cas d'une classification binaire. Donner, en fonction du contenu de la matrice de confusion les expressions de l'$accuracy$, la $precision$, le $recall$ et le $F_1 score$.


1. def: une matrice de confusion permet d'identitifié la validité des réponses d'un modele par rapport a la realité.

2.  M[0,0]= true positif ; M[1,1]= true negatif ; M[1,0]= false negatif ; M[0,1]= false positif.

    accuracy = (M[0,0]+M[1,1])/(M[0,0]+M[0,1]+M[1,0]+M[1,1]).$$
    precision=  M[0,0]/(M[0,0]+M[0,1]).$$
    recall/sensibily = M[0,0]/(M[0,0]+M[1,0]).$$
    F1-score= M[0,0]/ (M[0,0] + (M[1,0]+M[0,1])/2).$$

### Exercice 2.

On dispose de données (images IRM) relatives à des personnes atteintes (ou pas) de cancer. On entraîne alors un modèle pour détecter d'éventuelles cellules cancereuses sur les images. Celles-ci sont alors classifiées $1$ pour les images positives (patient atteint de cancer) et $0$ pour les images négatives.

On a ensuite testé le modèle sur des données d'autres patients pour lesquels le diagnostic est connu. Le résultat peut être récupéré à l'adresse 

<code>https://www.labri.fr/perso/zemmari/datasets/y_test_pred.csv</code>

#### Question 1.

Ecrire le code nécessaire pour charger le contenu du fichier dans un <code>DataFrame y</code>. On peut récupérer les premières lignes de <code> y </code> avec <code>y.head()</code>

In [11]:
import pandas as pd
data= pd.read_csv("https://www.labri.fr/perso/zemmari/datasets/y_test_pred.csv")
print(data)
#y=data.head()
y=data

     vt  vp
0     1   1
1     0   1
2     0   1
3     1   1
4     1   1
..   ..  ..
508   0   1
509   1   1
510   1   1
511   0   1
512   0   1

[513 rows x 2 columns]


Tester le code suivant pour apprendre comment accéder à une colonne d'un DataFrame et la transformer en numpy array.

In [8]:
y_vt = y["vt"]
y_vp = y["vp"]
y_vt

0      1
1      0
2      0
3      1
4      1
      ..
508    0
509    1
510    1
511    0
512    0
Name: vt, Length: 513, dtype: int64

In [9]:
import numpy as np
y_vt= np.array(y_vt)


#### Question 2.
Ecrire une fonction <code>compute_confusion_matrix(y_vt, y_vp)</code> permettant de calculer la matrice de confusion entre la vérité terrain <code>y_vt</code> et les prédictions <code>y_vp</code> (données dans le format pandas). La matrice de confusion calculée sera un numpy array 2x2.

*Indication :*  la fonction <code>logical_and(exp1, exp2)</code> du package <code>numpy</code> retourne <code>1</code> si les deux expressions sont vraies.


In [21]:
def compute_confusion_matrix(y_vt, y_vp):
    MatriceConfusion=np.zeros((2, 2))
    MatriceConfusion[0,0]= np.sum(np.logical_and(y_vt==0, y_vp==0))
    MatriceConfusion[0,1]= np.sum(np.logical_and(y_vt==0, y_vp==1))
    MatriceConfusion[1,0]= np.sum(np.logical_and(y_vt==1, y_vp==0))
    MatriceConfusion[1,1]= np.sum(np.logical_and(y_vt==1, y_vp==1))
    return MatriceConfusion

Calculer la matrice de confusions des données ci-dessus

In [22]:
newMatrice=compute_confusion_matrix(y_vt, y_vp)

def printMatrice(M):
    for i in range(M.shape[0]):
        for j in range(M.shape[1]):
            print("M[",i,",",j,"]=",M[i,j])
            
printMatrice(newMatrice)

M[ 0 , 0 ]= 3.0
M[ 0 , 1 ]= 190.0
M[ 1 , 0 ]= 3.0
M[ 1 , 1 ]= 317.0


#### Question 3.
Ecrire une fonction <code>compute_metrics(confusion_matrix)</code> permettant de calculer $accuracy$, la $precision$, le $recall$ et le $F_1 score$ à partir de la matrice de confusion donnée sous la forme d'un numpy array 2x2.

In [26]:
def compute_metrics(M):
    TN=M[0,0]
    FP=M[0,1]
    FN=M[1,0]
    TP=M[1,1]
    accuracy = (TP+TN)/(TP+FN+FP+TN)
    precision=  TP/(TP+FP)
    recall = TP/(TP+FN)
    F1_score= TP/ (TP + (FP+FN)/2)
    return accuracy,precision,recall,F1_score

Utiliser la fonction pour évaluer le modèle. Qu'en pensez-vous ? Donner une interprétation de chacune des métriques.

In [27]:
accuracy, precision, recall, F1_score = compute_metrics(newMatrice)

print("Précision:", precision)
print("Recall:", recall)
print("Accuracy:", accuracy)
print("F1-score:", f1_score)

Précision: 0.6252465483234714
Recall: 0.990625
Accuracy: 0.6237816764132553
F1-score: 0.03015075376884422


Précision (Precision) : La précision mesure la proportion des prédictions positives correctes parmi toutes les prédictions positives. Une précision de 0.625 signifie que 62,5 % des prédictions positives du modèle étaient correctes.

Rappel (Recall) : Le rappel mesure la proportion des vraies positives correctement prédites parmi toutes les vraies positives réelles. Un rappel de 0.990625 signifie que le modèle a correctement identifié 99,06 % des cas réellement positifs.

Exactitude (Accuracy) : L'exactitude mesure la proportion de toutes les prédictions correctes parmi toutes les prédictions effectuées par le modèle. Une exactitude de 0.6238 signifie que le modèle a correctement classé environ 62,38 % de toutes les instances.

F1-Score : Le score F1 est une mesure qui combine à la fois la précision et le rappel en une seule métrique. Un score F1 de 0.0302 indique que le modèle atteint un équilibre relativement faible entre la précision et le rappel.


En résumé, les résultats montrent que votre modèle a un bon rappel, ce qui signifie qu'il est capable de capturer la plupart des vraies positives, mais il a une précision relativement faible, ce qui signifie qu'il fait également beaucoup de fausses prédictions positives. L'exactitude est également faible, ce qui peut indiquer un déséquilibre de classe ou d'autres problèmes dans les données ou le modèle. Le score F1 est faible, suggérant que l'équilibre entre précision et rappel n'est pas optimal pour votre cas d'utilisation. Vous devrez peut-être ajuster votre modèle ou vos données pour améliorer ces métriques en fonction de vos objectifs spécifiques.
